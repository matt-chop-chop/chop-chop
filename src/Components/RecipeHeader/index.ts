export { default as RecipeHeader } from "./RecipeHeader";
