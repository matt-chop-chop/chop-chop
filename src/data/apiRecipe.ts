export const apiRecipe: ApiRecipe = {
  idMeal: "52922",
  strMeal: "Prawn & Fennel Bisque",
  strDrinkAlternate: null,
  strCategory: "Side",
  strArea: "French",
  strInstructions:
    "Shell the prawns, then fry the shells in the oil in a large pan for about 5 mins. Add the onion, fennel and carrots and cook for about 10 mins until the veg start to soften. Pour in the wine and brandy, bubble hard for about 1 min to drive off the alcohol, then add the tomatoes, stock and paprika. Cover and simmer for 30 mins. Meanwhile, chop the prawns.\r\nBlitz the soup as finely as you can with a stick blender or food processor, then press through a sieve into a bowl. Spend a bit of time really working the mixture through the sieve as this will give the soup its velvety texture.\r\nTip back into a clean pan, add the prawns and cook for 10 mins, then blitz again until smooth. You can make and chill this a day ahead or freeze it for 1 month. Thaw ovenight in the fridge. To serve, gently reheat in a pan with the cream. If garnishing, cook the 8 prawns in a little butter. Spoon into small bowls and top with the prawns and snipped fennel fronds.",
  strMealThumb:
    "https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg",
  strTags: "Soup,Warm,Seafood,Shellfish",
  strYoutube: "https://www.youtube.com/watch?v=4thpMbDakgM",
  strIngredient1: "Tiger Prawns",
  strIngredient2: "Olive Oil",
  strIngredient3: "Onion",
  strIngredient4: "Fennel",
  strIngredient5: "Carrots",
  strIngredient6: "Dry White Wine",
  strIngredient7: "Brandy",
  strIngredient8: "Chopped Tomatoes",
  strIngredient9: "Fish Stock",
  strIngredient10: "Paprika",
  strIngredient11: "Double Cream",
  strIngredient12: "Prawns",
  strIngredient13: "",
  strIngredient14: "",
  strIngredient15: "",
  strIngredient16: "",
  strIngredient17: "",
  strIngredient18: "",
  strIngredient19: "",
  strIngredient20: "",
  strMeasure1: "450g",
  strMeasure2: "4 tbs",
  strMeasure3: "1 large",
  strMeasure4: "1 large",
  strMeasure5: "2 chopped",
  strMeasure6: "150ml",
  strMeasure7: "1 tbs",
  strMeasure8: "400g",
  strMeasure9: "1L",
  strMeasure10: "2 pinches",
  strMeasure11: "150ml",
  strMeasure12: "8",
  strMeasure13: "",
  strMeasure14: "",
  strMeasure15: "",
  strMeasure16: "",
  strMeasure17: "",
  strMeasure18: "",
  strMeasure19: "",
  strMeasure20: "",
  strSource:
    "https://www.bbcgoodfood.com/recipes/71600/prawn-and-fennel-bisque",
  strImageSource: null,
  strCreativeCommonsConfirmed: null,
  dateModified: null,
};
