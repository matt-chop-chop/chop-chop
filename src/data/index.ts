export { recipe } from "./recipe";
export { apiRecipe } from "./apiRecipe";
