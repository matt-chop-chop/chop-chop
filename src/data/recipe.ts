export const recipe: Recipe = {
  id: "52922",
  area: "French",
  category: "Side",
  image: "https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg",
  ingredients: [
    {
      name: "Tiger Prawns",
      measurement: "450g",
      image: "https://www.themealdb.com/images/ingredients/Tiger Prawns.png",
    },
    {
      name: "Olive Oil",
      measurement: "4 tbs",
      image: "https://www.themealdb.com/images/ingredients/Olive Oil.png",
    },
    {
      name: "Onion",
      measurement: "1 large",
      image: "https://www.themealdb.com/images/ingredients/Onion.png",
    },
    {
      name: "Fennel",
      measurement: "1 large",
      image: "https://www.themealdb.com/images/ingredients/Fennel.png",
    },
    {
      name: "Carrots",
      measurement: "2 chopped",
      image: "https://www.themealdb.com/images/ingredients/Carrots.png",
    },
    {
      name: "Dry White Wine",
      measurement: "150ml",
      image: "https://www.themealdb.com/images/ingredients/Dry White Wine.png",
    },
    {
      name: "Brandy",
      measurement: "1 tbs",
      image: "https://www.themealdb.com/images/ingredients/Brandy.png",
    },
    {
      name: "Chopped Tomatoes",
      measurement: "400g",
      image:
        "https://www.themealdb.com/images/ingredients/Chopped Tomatoes.png",
    },
    {
      name: "Fish Stock",
      measurement: "1L",
      image: "https://www.themealdb.com/images/ingredients/Fish Stock.png",
    },
    {
      name: "Paprika",
      measurement: "2 pinches",
      image: "https://www.themealdb.com/images/ingredients/Paprika.png",
    },
    {
      name: "Double Cream",
      measurement: "150ml",
      image: "https://www.themealdb.com/images/ingredients/Double Cream.png",
    },
    {
      name: "Prawns",
      measurement: "8",
      image: "https://www.themealdb.com/images/ingredients/Prawns.png",
    },
  ],
  instructions: [
    "Shell the prawns, then fry the shells in the oil in a large pan for about 5 mins. ",
    "Add the onion, fennel and carrots and cook for about 10 mins until the veg start to soften. ",
    "Pour in the wine and brandy, bubble hard for about 1 min to drive off the alcohol, then add the tomatoes, stock and paprika. ",
    "Cover and simmer for 30 mins. ",
    "Meanwhile, chop the prawns.\r\nBlitz the soup as finely as you can with a stick blender or food processor, then press through a sieve into a bowl. ",
    "Spend a bit of time really working the mixture through the sieve as this will give the soup its velvety texture.\r\nTip back into a clean pan, add the prawns and cook for 10 mins, then blitz again until smooth. ",
    "You can make and chill this a day ahead or freeze it for 1 month. ",
    "Thaw ovenight in the fridge. ",
    "To serve, gently reheat in a pan with the cream. ",
    "If garnishing, cook the 8 prawns in a little butter. ",
    "Spoon into small bowls and top with the prawns and snipped fennel fronds.",
  ],
  name: "Prawn & Fennel Bisque",
  thumbnail:
    "https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg/preview",
  source: "https://www.bbcgoodfood.com/recipes/71600/prawn-and-fennel-bisque",
  tags: ["Soup", "Warm", "Seafood", "Shellfish"],
  ingredientCount: 12,
  time: 75,
  youtube: "https://www.youtube.com/watch?v=4thpMbDakgM",
};
