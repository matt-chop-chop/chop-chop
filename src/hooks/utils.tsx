export const getReactQueryError = (error?: unknown): Error | undefined => {
  if (!error) return;

  if (error instanceof Error) return error;

  if (typeof error === "string") return new Error(error);

  return new Error(`${error}`);
};

const parseIngredients = (apiRecipe: ApiRecipe): RecipeIngredient[] => {
  const ingredientKeys = Object.keys(apiRecipe).filter((key) =>
    key.includes("strIngredient")
  );

  const measurementKeys = Object.keys(apiRecipe).filter((key) =>
    key.includes("strMeasure")
  );

  const _ingredients = ingredientKeys.map((key, index) => {
    return {
      name: apiRecipe[key as keyof ApiRecipe] || "",
      measurement: apiRecipe[measurementKeys[index] as keyof ApiRecipe] || "",
      image: `https://www.themealdb.com/images/ingredients/${
        apiRecipe[key as keyof ApiRecipe]
      }.png`,
    };
  });

  const ingredients = _ingredients.filter(
    (ingredient) => ingredient.name !== ""
  );

  return ingredients;
};

const parseInstructions = (instructions: string): string[] => {
  const instructionsWithDelimiter =
    instructions.replaceAll(". ", ". @£$") || "";
  const instructionSteps = instructionsWithDelimiter.split("@£$");
  return instructionSteps;
};

const getSeconds = (instructions: string): number => {
  const secondRegEx =
    "[0-9]+(?=sec)|[0-9]+(?= sec)|[0-9]+(?=second)|[0-9]+(?= second)|[0-9]+(?=secs)|[0-9]+(?= secs)|[0-9]+(?=seconds)|[0-9]+(?= seconds)";
  const matches = instructions.match(new RegExp(secondRegEx, "g"));
  const matchesAsNumbers = matches?.map((match) => parseInt(match)) || [];
  const sum = matchesAsNumbers.reduce((a, b) => a + b, 0);
  return sum;
};

const getMinutes = (instructions: string): number => {
  const minuteRegEx =
    "[0-9]+(?=min)|[0-9]+(?= min)|[0-9]+(?=minute)|[0-9]+(?= minute)|[0-9]+(?=mins)|[0-9]+(?= mins)|[0-9]+(?=minutes)|[0-9]+(?= minutes)";
  const matches = instructions.match(new RegExp(minuteRegEx, "g"));
  const matchesAsNumbers = matches?.map((match) => parseInt(match)) || [];
  const sum = matchesAsNumbers.reduce((a, b) => a + b, 0);
  return sum * 60;
};

const getHours = (instructions: string): number => {
  const hourRegEx =
    "[0-9]+(?=hr)|[0-9]+(?= hr)|[0-9]+(?=hour)|[0-9]+(?= hour)|[0-9]+(?=hrs)|[0-9]+(?= hrs)|[0-9]+(?=hours)|[0-9]+(?= hours)";
  const matches = instructions.match(new RegExp(hourRegEx, "g"));
  const matchesAsNumbers = matches?.map((match) => parseInt(match)) || [];
  const sum = matchesAsNumbers.reduce((a, b) => a + b, 0);
  return sum * 3600;
};

const parseCookingTime = (instructions: string): number => {
  const _instructions = instructions.toLocaleLowerCase();

  const timeAsSeconds =
    getSeconds(_instructions) +
    getMinutes(_instructions) +
    getHours(_instructions);
  const timeAsMinutes = timeAsSeconds / 60;

  // rounded up to the nearest 5 minutes
  const roundedTimeAsMinutes = Math.ceil(timeAsMinutes / 5) * 5;

  // add on a contingency 15 minutes
  return roundedTimeAsMinutes + 15;
};

export const convertApiRecipeToRecipe = (
  apiRecipe: ApiRecipe | null
): Recipe | null => {
  if (!apiRecipe) return null;

  const ingredients = parseIngredients(apiRecipe);
  const tags = apiRecipe?.strTags ? apiRecipe.strTags.split(",") : [];
  const instructions = parseInstructions(apiRecipe?.strInstructions || "");

  return {
    id: apiRecipe?.idMeal || "",
    area: apiRecipe?.strArea || "",
    category: apiRecipe?.strCategory || "",
    image: apiRecipe?.strMealThumb || "",
    ingredients,
    instructions,
    name: apiRecipe?.strMeal || "",
    thumbnail: apiRecipe?.strMealThumb
      ? apiRecipe.strMealThumb + "/preview"
      : "",
    source: apiRecipe?.strSource || "",
    tags,
    ingredientCount: ingredients.length,
    time: parseCookingTime(apiRecipe?.strInstructions || ""),
    youtube: apiRecipe?.strYoutube || "",
  };
};

export const convertApiRecipesToRecipes = (
  apiRecipes: (ApiRecipe | null)[]
): (Recipe | null)[] => {
  return apiRecipes.map((apiRecipe) => convertApiRecipeToRecipe(apiRecipe));
};

export const convertApiCategoriesToCategories = (
  apiCategories: ApiCategory[]
): { name: string }[] => {
  return apiCategories.map((apiCategory) => {
    return { name: apiCategory?.strCategory || "" };
  });
};

export const convertApiAreasToAreas = (
  apiAreas: ApiArea[]
): { name: string }[] => {
  return apiAreas.map((apiArea) => {
    return { name: apiArea?.strArea || "" };
  });
};

export const convertApiIngredientsToIngredients = (
  apiIngredients: ApiIngredient[]
): SelectableIngredient[] => {
  return apiIngredients.map((apiIngredient) => {
    return {
      name: apiIngredient?.strIngredient || "",
      description: apiIngredient?.strDescription || "",
    };
  });
};
