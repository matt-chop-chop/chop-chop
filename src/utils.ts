export const fromatCookingTime = (time: number): string => {
  if (time <= 60) return `${time} minutes`;

  if (time < 120) return `${Math.floor(time / 60)} hour ${time % 60} minutes`;

  return `${Math.floor(time / 60)} hours ${time % 60} minutes`;
};
