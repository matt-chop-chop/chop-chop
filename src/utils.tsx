export const fromatCookingTime = (time: number): string => {
  if (time < 60) return `${time} minutes`;

  if (time < 120) return `${Math.floor(time / 60)} hour ${time % 60} minutes`;

  return `${Math.floor(time / 60)} hours ${time % 60} minutes`;
};

export const parseRecipe = (recipe: Recipe): string => {
  let response =
    recipe.name +
    "\n" +
    `Cooking Time: ${fromatCookingTime(recipe.time)}` +
    "\n" +
    `Ingredient Count: ${recipe.ingredientCount}` +
    "\n" +
    `Meal Type: ${recipe.category}` +
    "\n" +
    `Place of Origin: ${recipe.area}` +
    "\n" +
    "Tags:";
  recipe.tags.map((tag) => response + " " + tag);
  response += "\n" + "Ingredients:";
  recipe.ingredients.map(
    (ingredient) =>
      (response += "\n" + `â€¢ ${ingredient.measurement} ${ingredient.name}`)
  );
  response += "\n" + "Method:";
  recipe.instructions.map(
    (instruction) => (response += "\n" + `- ${instruction}`)
  );
  response +=
    "\n" +
    `Youtube Tutorial: ${recipe.youtube}` +
    "\n" +
    `Recipe Source: ${recipe.source}`;
  return response;
};

export const download = (recipe: Recipe) => {
  const link = document.createElement("a");
  link.setAttribute(
    "href",
    "data:text/plain;charset=utf-8," + encodeURIComponent(parseRecipe(recipe))
  );
  link.setAttribute("download", `${recipe.name}.txt`);

  link.style.display = "none";
  document.body.appendChild(link);

  link.click();

  document.body.removeChild(link);
};
