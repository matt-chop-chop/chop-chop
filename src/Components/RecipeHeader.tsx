import { download } from "@/utils";
import { DownloadIcon } from "@chakra-ui/icons";
import { Flex, Heading } from "@chakra-ui/react";
import { RecipeDetailsGrid, RecipeImage, RecipeTags } from "@/Components";

type RecipeHeaderProps = {
  recipe: Recipe;
};

const RecipeHeader = ({ recipe }: RecipeHeaderProps) => {
  return (
    <Flex
      borderBottom="2px solid var(--chakra-colors-light-emphasis)"
      flexDirection="column"
      p={4}
    >
      <RecipeImage image={recipe.image} />
      <Flex alignItems="center" justifyContent="space-between" width="100%">
        <Heading
          as="h1"
          mt={6}
          textDecoration="underline"
          variant="heading-large"
        >
          {recipe.name}
        </Heading>
        <DownloadIcon
          boxSize={5}
          cursor="pointer"
          onClick={() => download(recipe)}
          transition="transform 250ms ease-out"
          _hover={{
            transition: "transform 250ms ease-in",
            transform: "scale(1.5)",
          }}
        />
      </Flex>
      <RecipeDetailsGrid
        area={recipe.area}
        category={recipe.category}
        ingredientCount={recipe.ingredientCount}
        time={recipe.time}
      />
      <RecipeTags tags={recipe.tags} />
    </Flex>
  );
};

export default RecipeHeader;
