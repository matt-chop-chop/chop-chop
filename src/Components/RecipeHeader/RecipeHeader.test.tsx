import { shallow } from "enzyme";
import React from "react";
import RecipeHeader, { RecipeHeaderProps } from "./RecipeHeader";
import { download } from "@/utils";
import { recipe } from "@/data/recipe";

jest.mock("@/utils", () => ({
  download: jest.fn(),
}));

const props: RecipeHeaderProps = {
  recipe: recipe,
};

describe("RecipeHeader", () => {
  it("is exported from RecipeHeader", () => {
    expect(RecipeHeader).toBeDefined();
  });

  it("renders as expected ", () => {
    const wrapper = shallow(<RecipeHeader {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("calls download when the DownloadIcon is clicked ", () => {
    const wrapper = shallow(<RecipeHeader {...props} />);

    const DownloadIcon = wrapper.find("DownloadIcon");
    expect(DownloadIcon).toHaveLength(1);

    DownloadIcon.simulate("click");

    expect(download).toHaveBeenCalledWith(props.recipe);
  });
});
