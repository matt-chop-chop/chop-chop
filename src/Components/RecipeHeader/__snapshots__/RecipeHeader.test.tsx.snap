// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeHeader renders as expected  1`] = `
<Flex
  borderBottom="2px solid var(--chakra-colors-light-text)"
  flexDirection="column"
  p={4}
>
  <RecipeImage
    image="https://www.themealdb.com/images/media/meals/rtwwvv1511799504.jpg"
  />
  <Flex
    alignItems="center"
    justifyContent="space-between"
    width="100%"
  >
    <Heading
      as="h1"
      mt={6}
      textDecoration="underline"
      variant="heading-large"
    >
      Prawn & Fennel Bisque
    </Heading>
    <DownloadIcon
      _hover={
        {
          "transform": "scale(1.5)",
          "transition": "transform 250ms ease-in",
        }
      }
      boxSize={5}
      cursor="pointer"
      onClick={[Function]}
      transition="transform 250ms ease-out"
    />
  </Flex>
  <RecipeDetailsGrid
    area="French"
    category="Side"
    ingredientCount={12}
    time={75}
  />
  <RecipeTags
    tags={
      [
        "Soup",
        "Warm",
        "Seafood",
        "Shellfish",
      ]
    }
  />
</Flex>
`;
